File: tools/bridge/bridge_sql.py
"""
bridge_sql.py — canonical bridge client for troy-sql-executor / run_sql RPC.

Replaces the legacy wrapper that masked pg errors as `sql_error` / `command:null`
and silently returned `rows:[]`.

Contract:
  ok  -> returns {"rows": [...], "command": "<UPPER>", "rowcount": int, "raw": {...}}
  bad -> raises BridgeSqlError carrying http_status, code, message, details, hint, raw

Failure modes detected (all raise, none return):
  1. HTTP non-2xx                            -> http_status + body
  2. Bridge envelope error (statusCode!=200) -> upstream error
  3. PostgREST error body (code/message/...)  -> verbatim pg fields
  4. Lambda body has `error`                 -> raised
  5. Lambda body has `sql_error`             -> raised (legacy mask)
  6. `command` is None / "" / missing        -> raised (NEW: was silent success)
  7. `rows` missing entirely                 -> raised (suspicious shape)

Usage:
    from bridge_sql import run_sql, BridgeSqlError
    try:
        out = run_sql("SELECT 1 AS x;")
        print(out["rows"])
    except BridgeSqlError as e:
        print(e.as_dict())  # full pg context
"""

from __future__ import annotations

import json
import os
import socket
import ssl
import time
import urllib.error
import urllib.request
from dataclasses import dataclass, field, asdict
from typing import Any

BRIDGE_URL_DEFAULT = (
    "https://zdgnab3py0.execute-api.ap-southeast-2.amazonaws.com/prod/lambda/invoke"
)


# ---------- error type --------------------------------------------------------

@dataclass
class BridgeSqlError(Exception):
    """Raised on any non-OK bridge/SQL response. Carries full pg context."""
    http_status: int | None = None
    code: str | None = None            # PostgREST/pg SQLSTATE-ish code
    message: str | None = None         # primary human message
    details: str | None = None         # pg detail
    hint: str | None = None            # pg hint
    where: str | None = None           # which check tripped
    sql: str | None = None             # SQL we sent (truncated)
    raw: dict | None = field(default=None, repr=False)

    def __post_init__(self) -> None:
        # Build a useful __str__ for callers that just print(e)
        parts = [f"BridgeSqlError[{self.where or '?'}]"]
        if self.http_status is not None:
            parts.append(f"http={self.http_status}")
        if self.code:
            parts.append(f"code={self.code}")
        if self.message:
            parts.append(f"msg={self.message}")
        if self.details:
            parts.append(f"details={self.details}")
        if self.hint:
            parts.append(f"hint={self.hint}")
        super().__init__(" | ".join(parts))

    def as_dict(self) -> dict:
        return {k: v for k, v in asdict(self).items() if v is not None}


# ---------- helpers -----------------------------------------------------------

_PG_FIELDS = ("code", "message", "details", "hint")


def _extract_pg(obj: Any) -> dict[str, Any]:
    """Pull standard pg/PostgREST fields from any dict-shaped body."""
    if not isinstance(obj, dict):
        return {}
    return {k: obj.get(k) for k in _PG_FIELDS if obj.get(k) is not None}


def _truncate(s: str, n: int = 240) -> str:
    return s if len(s) <= n else s[:n] + "...[truncated]"


# ---------- main entrypoint ---------------------------------------------------

def run_sql(
    sql: str,
    *,
    bridge_url: str | None = None,
    api_key: str | None = None,
    timeout: float = 30.0,
    fn: str = "troy-sql-executor",
) -> dict:
    """
    Invoke troy-sql-executor via the bridge with fail-fast semantics.

    Returns: {"rows": list, "command": str, "rowcount": int, "raw": dict}
    Raises:  BridgeSqlError on ANY failure path (incl. command:null).
    """
    bridge_url = bridge_url or os.environ.get("BRIDGE_URL", BRIDGE_URL_DEFAULT)
    api_key = api_key or os.environ.get("BRIDGE_API_KEY", "")
    if not api_key:
        raise BridgeSqlError(
            where="config",
            message="BRIDGE_API_KEY not set (env BRIDGE_API_KEY or pass api_key=)",
        )

    # Bridge envelope: troy-sql-executor uses NESTED payload (sql under payload)
    envelope = {"fn": fn, "payload": {"sql": sql}}
    body = json.dumps(envelope).encode("utf-8")

    req = urllib.request.Request(
        bridge_url,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "x-api-key": api_key,
        },
    )

    # ---- transport ---------------------------------------------------------
    raw_text: str = ""
    http_status: int | None = None
    last_err: Exception | None = None

    # Single retry for the known sandbox proxy "DNS cache overflow" symptom.
    for attempt in (1, 2):
        try:
            opener = urllib.request.build_opener()
            opener.addheaders = [("Connection", "keep-alive")]
            with opener.open(req, timeout=timeout) as r:
                http_status = r.status
                raw_text = r.read().decode("utf-8", errors="replace")
            break
        except urllib.error.HTTPError as e:
            http_status = e.code
            raw_text = e.read().decode("utf-8", errors="replace") if e.fp else ""
            # Try to surface pg error from non-2xx body too
            try:
                parsed = json.loads(raw_text) if raw_text else {}
            except Exception:
                parsed = {}
            pg = _extract_pg(parsed)
            raise BridgeSqlError(
                http_status=http_status,
                where="http",
                message=pg.get("message") or f"HTTP {http_status}",
                code=pg.get("code"),
                details=pg.get("details"),
                hint=pg.get("hint"),
                sql=_truncate(sql),
                raw={"body": raw_text[:4000]},
            ) from e
        except (urllib.error.URLError, socket.timeout, ssl.SSLError, ConnectionError) as e:
            last_err = e
            if attempt == 1:
                time.sleep(0.5)
                continue
            raise BridgeSqlError(
                where="transport",
                message=f"{type(e).__name__}: {e}",
                sql=_truncate(sql),
            ) from e

    # ---- parse outer ------------------------------------------------------
    try:
        outer = json.loads(raw_text) if raw_text else {}
    except json.JSONDecodeError as e:
        raise BridgeSqlError(
            http_status=http_status,
            where="parse_outer",
            message=f"non-JSON bridge response: {e}",
            raw={"body": raw_text[:4000]},
            sql=_truncate(sql),
        ) from e

    # API Gateway wraps Lambda response as {statusCode, body:"<json string>"}
    inner: Any = outer
    if isinstance(outer, dict) and "statusCode" in outer and "body" in outer:
        if outer.get("statusCode") and outer["statusCode"] >= 400:
            pg = _extract_pg(outer)
            raise BridgeSqlError(
                http_status=outer["statusCode"],
                where="bridge_envelope",
                message=pg.get("message") or f"bridge statusCode={outer['statusCode']}",
                code=pg.get("code"),
                details=pg.get("details"),
                hint=pg.get("hint"),
                raw=outer,
                sql=_truncate(sql),
            )
        try:
            inner = json.loads(outer["body"]) if isinstance(outer["body"], str) else outer["body"]
        except json.JSONDecodeError as e:
            raise BridgeSqlError(
                http_status=http_status,
                where="parse_inner",
                message=f"non-JSON Lambda body: {e}",
                raw=outer,
                sql=_truncate(sql),
            ) from e

    if not isinstance(inner, dict):
        raise BridgeSqlError(
            http_status=http_status,
            where="shape",
            message="Lambda body is not an object",
            raw={"body": inner},
            sql=_truncate(sql),
        )

    # ---- legacy / explicit error fields -----------------------------------
    # `error` is generic; `sql_error` is the legacy mask used by troy-sql-executor.
    explicit = inner.get("error") or inner.get("sql_error")
    if explicit:
        pg = _extract_pg(inner)
        # explicit may itself be a dict carrying pg fields
        if isinstance(explicit, dict):
            pg = {**_extract_pg(explicit), **pg}
            msg = explicit.get("message") or json.dumps(explicit)
        else:
            msg = str(explicit)
        raise BridgeSqlError(
            http_status=http_status,
            where="lambda_error",
            message=pg.get("message") or msg,
            code=pg.get("code"),
            details=pg.get("details"),
            hint=pg.get("hint"),
            raw=inner,
            sql=_truncate(sql),
        )

    # ---- THE FIX: command:null is FAILURE, not success --------------------
    cmd = inner.get("command")
    if cmd in (None, "", "null"):
        pg = _extract_pg(inner)
        raise BridgeSqlError(
            http_status=http_status,
            where="command_null",
            message=pg.get("message") or "command is null (pg execution did not produce a tag)",
            code=pg.get("code"),
            details=pg.get("details"),
            hint=pg.get("hint") or "split multi-statement SQL; troy-sql-executor does not support BEGIN/COMMIT chains",
            raw=inner,
            sql=_truncate(sql),
        )

    # ---- shape sanity -----------------------------------------------------
    if "rows" not in inner:
        raise BridgeSqlError(
            http_status=http_status,
            where="shape",
            message="response missing `rows` field",
            raw=inner,
            sql=_truncate(sql),
        )

    rows = inner["rows"] or []
    return {
        "rows": rows,
        "command": str(cmd).upper(),
        "rowcount": inner.get("rowcount", len(rows) if isinstance(rows, list) else 0),
        "raw": inner,
    }


# ---------- self-test ---------------------------------------------------------

def _selftest() -> int:
    """Run a small matrix. Requires BRIDGE_API_KEY in env."""
    cases: list[tuple[str, str, str]] = [
        ("ok_select",       "SELECT 1 AS x;",                                   "ok"),
        ("ok_no_semi",      "SELECT 2 AS x",                                    "ok"),
        ("bad_syntax",      "SELEKT 1;",                                        "raise"),
        ("bad_relation",    "SELECT * FROM no_such_table_xyz_42;",              "raise"),
        ("multi_stmt",      "SELECT 1; SELECT 2;",                              "raise"),  # not supported
        ("empty",           "",                                                 "raise"),
    ]
    failed = 0
    for name, sql, expect in cases:
        try:
            out = run_sql(sql)
            got = "ok"
            print(f"[{name}] -> ok rows={len(out['rows'])} cmd={out['command']}")
        except BridgeSqlError as e:
            got = "raise"
            print(f"[{name}] -> raise {e}")
        if got != expect:
            print(f"  !! expected {expect} got {got}")
            failed += 1
    print(f"\nfailed={failed}/{len(cases)}")
    return failed


if __name__ == "__main__":
    raise SystemExit(_selftest())

File: tools/bridge/bridgeSql.mjs
// bridgeSql.mjs — canonical bridge client for troy-sql-executor.
// Fail-fast on command:null, non-OK HTTP, and PostgREST/pg error envelopes.
// Mirrors bridge_sql.py contract.

const BRIDGE_URL_DEFAULT =
  "https://zdgnab3py0.execute-api.ap-southeast-2.amazonaws.com/prod/lambda/invoke";

const PG_FIELDS = ["code", "message", "details", "hint"];

export class BridgeSqlError extends Error {
  constructor({ where, http_status, code, message, details, hint, sql, raw }) {
    const tag = `BridgeSqlError[${where || "?"}]`;
    const parts = [tag];
    if (http_status != null) parts.push(`http=${http_status}`);
    if (code) parts.push(`code=${code}`);
    if (message) parts.push(`msg=${message}`);
    if (details) parts.push(`details=${details}`);
    if (hint) parts.push(`hint=${hint}`);
    super(parts.join(" | "));
    this.name = "BridgeSqlError";
    this.where = where;
    this.http_status = http_status;
    this.code = code;
    this.pgMessage = message;
    this.details = details;
    this.hint = hint;
    this.sql = sql;
    this.raw = raw;
  }
  toJSON() {
    return {
      name: this.name,
      where: this.where,
      http_status: this.http_status,
      code: this.code,
      message: this.pgMessage,
      details: this.details,
      hint: this.hint,
      sql: this.sql,
    };
  }
}

function pickPg(obj) {
  if (!obj || typeof obj !== "object") return {};
  const out = {};
  for (const f of PG_FIELDS) if (obj[f] != null) out[f] = obj[f];
  return out;
}

function trunc(s, n = 240) {
  return s.length <= n ? s : s.slice(0, n) + "...[truncated]";
}

/**
 * runSql(sql, opts?) -> { rows, command, rowcount, raw }
 * Throws BridgeSqlError on any failure, including command:null.
 */
export async function runSql(sql, opts = {}) {
  const bridgeUrl = opts.bridgeUrl || process.env.BRIDGE_URL || BRIDGE_URL_DEFAULT;
  const apiKey = opts.apiKey || process.env.BRIDGE_API_KEY;
  const fn = opts.fn || "troy-sql-executor";
  const timeoutMs = opts.timeoutMs ?? 30000;

  if (!apiKey) {
    throw new BridgeSqlError({
      where: "config",
      message: "BRIDGE_API_KEY not set",
    });
  }

  const ctl = new AbortController();
  const timer = setTimeout(() => ctl.abort(), timeoutMs);

  let r, rawText;
  try {
    r = await fetch(bridgeUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json", "x-api-key": apiKey },
      body: JSON.stringify({ fn, payload: { sql } }),
      signal: ctl.signal,
    });
    rawText = await r.text();
  } catch (e) {
    throw new BridgeSqlError({
      where: "transport",
      message: `${e.name}: ${e.message}`,
      sql: trunc(sql),
    });
  } finally {
    clearTimeout(timer);
  }

  // Non-2xx — try to surface pg fields from body
  if (!r.ok) {
    let parsed = {};
    try { parsed = JSON.parse(rawText); } catch { /* ignore */ }
    const pg = pickPg(parsed);
    throw new BridgeSqlError({
      where: "http",
      http_status: r.status,
      message: pg.message || `HTTP ${r.status}`,
      code: pg.code,
      details: pg.details,
      hint: pg.hint,
      sql: trunc(sql),
      raw: { body: rawText.slice(0, 4000) },
    });
  }

  let outer;
  try { outer = JSON.parse(rawText); }
  catch (e) {
    throw new BridgeSqlError({
      where: "parse_outer",
      http_status: r.status,
      message: `non-JSON bridge response: ${e.message}`,
      sql: trunc(sql),
      raw: { body: rawText.slice(0, 4000) },
    });
  }

  // Unwrap API Gateway envelope { statusCode, body }
  let inner = outer;
  if (outer && typeof outer === "object" && "statusCode" in outer && "body" in outer) {
    if (outer.statusCode >= 400) {
      const pg = pickPg(outer);
      throw new BridgeSqlError({
        where: "bridge_envelope",
        http_status: outer.statusCode,
        message: pg.message || `bridge statusCode=${outer.statusCode}`,
        code: pg.code, details: pg.details, hint: pg.hint,
        sql: trunc(sql), raw: outer,
      });
    }
    try {
      inner = typeof outer.body === "string" ? JSON.parse(outer.body) : outer.body;
    } catch (e) {
      throw new BridgeSqlError({
        where: "parse_inner",
        message: `non-JSON Lambda body: ${e.message}`,
        sql: trunc(sql), raw: outer,
      });
    }
  }

  if (!inner || typeof inner !== "object") {
    throw new BridgeSqlError({
      where: "shape",
      message: "Lambda body is not an object",
      sql: trunc(sql), raw: { body: inner },
    });
  }

  // Explicit error fields (incl. legacy `sql_error` mask)
  const explicit = inner.error ?? inner.sql_error;
  if (explicit) {
    const pg = typeof explicit === "object"
      ? { ...pickPg(explicit), ...pickPg(inner) }
      : pickPg(inner);
    const msg = typeof explicit === "object"
      ? (explicit.message || JSON.stringify(explicit))
      : String(explicit);
    throw new BridgeSqlError({
      where: "lambda_error",
      message: pg.message || msg,
      code: pg.code, details: pg.details, hint: pg.hint,
      raw: inner, sql: trunc(sql),
    });
  }

  // THE FIX: command:null is failure
  const cmd = inner.command;
  if (cmd == null || cmd === "" || cmd === "null") {
    const pg = pickPg(inner);
    throw new BridgeSqlError({
      where: "command_null",
      message: pg.message || "command is null (pg execution did not produce a tag)",
      code: pg.code,
      details: pg.details,
      hint: pg.hint || "split multi-statement SQL; BEGIN/COMMIT chains not supported",
      raw: inner, sql: trunc(sql),
    });
  }

  if (!("rows" in inner)) {
    throw new BridgeSqlError({
      where: "shape",
      message: "response missing `rows` field",
      raw: inner, sql: trunc(sql),
    });
  }

  const rows = inner.rows ?? [];
  return {
    rows,
    command: String(cmd).toUpperCase(),
    rowcount: inner.rowcount ?? (Array.isArray(rows) ? rows.length : 0),
    raw: inner,
  };
}
